src/main/java/io/odpf/firehose/sink/mongodb/MongoSink.java

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * MongoDB sink.
+ */
+public class MongoSink extends AbstractSink {
+
+    private final MongoRequestHandler mongoRequestHandler;
+    private final List<WriteModel<Document>> bulkRequest = new ArrayList<>();
+    private final List<String> mongoRetryStatusCodeBlacklist;
+    private final MongoCollection<Document> mongoCollection;
+    private final MongoClient mongoClient;
+    private List<Message> messages;
+
+    /**
+     * Instantiates a new Mongo sink.
+     *
+     * @param instrumentation               the instrumentation
+     * @param sinkType                      the sink type
+     * @param mongoCollection               the mongo collection
+     * @param mongoClient                   the mongo client
+     * @param mongoRequestHandler           the mongo request handler
+     * @param mongoRetryStatusCodeBlacklist the mongo retry status code blacklist
+     */
+    public MongoSink(Instrumentation instrumentation, String sinkType, MongoCollection<Document> mongoCollection, MongoClient mongoClient, MongoRequestHandler mongoRequestHandler,
+                     List<String> mongoRetryStatusCodeBlacklist) {
+        super(instrumentation, sinkType);
+        this.mongoCollection = mongoCollection;
+        this.mongoRequestHandler = mongoRequestHandler;
+        this.mongoClient = mongoClient;
+        this.mongoRetryStatusCodeBlacklist = mongoRetryStatusCodeBlacklist;
+    }
+
+    @Override
+    protected void prepare(List<Message> messageList) {
+        this.messages = messageList;
+        bulkRequest.clear();
+
+        messages.forEach(message -> bulkRequest.add(mongoRequestHandler.getRequest(message)));
+    }
+
+    @Override
+    protected List<Message> execute() throws Exception {
+        List<BulkWriteError> bulkWriteErrors = processRequest();
+        if (!bulkWriteErrors.isEmpty()) {
+            getInstrumentation().logWarn("Bulk request failed");
+
+            handleBulkWriteErrors(bulkWriteErrors);
+        }
+
+        List<Message> failedMessages = new ArrayList<>();
+
+        bulkWriteErrors.forEach((bulkWriteError) -> {
+            if (!mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode()))) {
+                failedMessages.add(messages.get(bulkWriteError.getIndex()));
+            }
+        });
+        return failedMessages;
+    }
+
+    @Override
+    public void close() throws IOException {
+        getInstrumentation().logInfo("MongoDB connection closing");
+        this.mongoClient.close();
+    }
+
+    /**
+     * Process request list.
+     *
+     * @return the list
+     */
+    List<BulkWriteError> processRequest() {
+        List<BulkWriteError> bulkWriteErrors = new ArrayList<>();
+        try {
+            handleBulkWriteResult(mongoCollection.bulkWrite(bulkRequest));
+        } catch (BulkWriteException bulkWriteException) {
+
+            bulkWriteErrors = bulkWriteException.getWriteErrors();
```

Comment:
avoid throwing exception here. It's expensive. Just return the errors as part of method.